text/0000-path-clarity.md

```diff
@@ -0,0 +1,734 @@
+
+In the survey data collected in both 2016 and 2017, learnability and ergonomics
+issues were one of the major challenges for people using or considering
+Rust. While there were other features that were raised more frequently than the
+module system (lifetimes for example), ideally the module system, which isn't
+*meant* to be novel, would not be a learnability problem at all!
+
+Here are some select quotes (these are not the only responses that mention the module
+system):
+
+> Also the module system is confusing (not that I say is wrong, just confusing
+> until you are experienced in it).
+
+> a colleague of mine that started rust got really confused over the module
+> system
+
+> You had to import everything in the main module, but you also had to in
+> submodules, but if it was only imported in a submodule it wouldn't work.
+
+> I especially find the modules and crates design weird and verbose
+
+> fix the module system
+
+One user states that the reason they stopped using Rust was that the
+"module system is really unintuitive." Similar data is present in the 2016 survey.
+
+Experiences along similar lines can be found in Rust forums, StackOverflow, and
+similar, some of which has been collected into [a gist][learning-modules].
+
+[learning-modules]: https://gist.github.com/aturon/2f10f19f084f39330cfe2ee028b2ea0c
+
+The problems presented above represent a boiled down subset of the problems
+raised in this feedback.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## As we would teach it
+
+The following sections sketch a plausible way of teaching the module system once
+this RFC has been fully implemented.
+
+### Using external dependencies
+
+To add an external dependency, record it in the `[dependencies]` section of
+`Cargo.toml`:
+
+```toml
+[dependencies]
+serde = "1.0.0"
+```
+
+By default, crates have an automatic dependency on `std`, the standard library.
+
+Once your dependency has been added, you can bring it or its exports into scope with
+`use` declarations:
+
+```rust
+use std; // bring `std` itself into scope
+use std::vec::Vec;
+
+use serde::Serialize;
+```
+
+Note that these `use` declarations all begin with a crate name.
+
+Once an item is in scope, you can reference it directly within definitions:
+
+```rust
+// Either of these work, because we brought `std` into scope:
+fn make_vec() -> Vec<u8> { ... }
+fn make_vec() -> std::vec::Vec<u8> { ... }
+
+// Only the first of these work, because we didn't bring `serde` into scope:
+impl Serialize for MyType { ... }
+impl serde::Serialize for MyType { ... } // the name `serde` is not in scope here
+```
+
+You can also reference items from a crate without bringing them into scope by
+writing a **fully qualified path**, designated by a leading `::`, as follows:
+
+```rust
+impl ::serde::Serialize for MyType { ... }
+```
+
+All `use` declarations are interpreted as fully qualified paths, making the
+leading `::` optional for them.
```

Comment:
Isn't this a breaking change? What if I have a module in my existing code called `serde`?

Reply:
Yes, the guide is describing how things will look under the *full implementation*, which means a future epoch.